### ISSUE TITLE
What columns does `group` pass through?

### ISSUE THREAD

**Issue by richb-hanover** [PRQL/prql#3490]
### What's up?

Part of the reason I was confused in #2947 is that it appears that the `group` transform doesn't retain its [key_columns](https://prql-lang.org/book/reference/stdlib/transforms/group.html) (or at least, I don't know how to retain them).

The example below joins a bunch of tables, with the aim of summarizing the names of the tracks for each city (and country).

Here's the output of that query (as-is) from the Playground:

name | ct1
-- | --
For Those About To Rock (We Salute You) | 1
Balls to the Wall | 2
Fast As a Shark | 1
Restless and Wild | 1

Note that `billing_city` and `billing_country` do not appear to be in the table passed to the second `group` transform

And un-commenting either `sort` or `take 1` leaves only the `ct1` column.

Is this intended behavior? Thanks.

```
from inv=invoices
join item=invoice_items (==invoice_id)
join t=tracks (item.track_id==t.track_id)

group {inv.billing_city, inv.billing_country} (

  group {t.name} (
    aggregate {
    ct1 = count t.name,
    }
  #sort {-ct1}
  #take 1
  )
)
```

**Comment by max-sixty** [comment]
Without the nested group, it does retain the columns:

```elm
from inv=invoices
join item=invoice_items (==invoice_id)
join t=tracks (item.track_id==t.track_id)

group {inv.billing_city, inv.billing_country, t.name} (
  aggregate (c=count this)
)
sort (-c)
take 10
```

```sql
SELECT
  inv.billing_city,
  inv.billing_country,
  t.name,
  COUNT(*) AS c
FROM
  invoices AS inv
  JOIN invoice_items AS item ON inv.invoice_id = item.invoice_id
  JOIN tracks AS t ON item.track_id = t.track_id
GROUP BY
  inv.billing_city,
  inv.billing_country,
  t.name
ORDER BY
  c DESC
LIMIT
  10

-- Generated by PRQL compiler version:0.9.4 (https://prql-lang.org)

```

```json
[
  {"billing_city": "London", "billing_country": "United Kingdom", "name": "Firmamento", "c": 2},
  {"billing_city": "Winnipeg", "billing_country": "Canada", "name": "Onde Você Mora?", "c": 2},
  {"billing_city": "Berlin", "billing_country": "Germany", "name": "The Number Of The Beast", "c": 2},
  {"billing_city": "Stockholm", "billing_country": "Sweden", "name": "Polly", "c": 2},
  {"billing_city": "São Paulo", "billing_country": "Brazil", "name": "Untitled", "c": 2},
  {"billing_city": "Helsinki", "billing_country": "Finland", "name": "Eruption", "c": 2},
  {"billing_city": "Berlin", "billing_country": "Germany", "name": "Eruption", "c": 2},
  {"billing_city": "Rome", "billing_country": "Italy", "name": "For Those About To Rock (We Salute You)", "c": 1},
  {"billing_city": "Yellowknife", "billing_country": "Canada", "name": "Balls to the Wall", "c": 1},
  {"billing_city": "Brasília", "billing_country": "Brazil", "name": "Fast As a Shark", "c": 1}
]
```

What is the role of the nested group?

I'd need to think more about how we should be handling nested groups — I'm not claiming it's handled well — but is it needed for this specific query?

**Comment by richb-hanover** [comment]
The query above (without the nested `group`) doesn't really answer my question, which is: _What's the most popular song for each city in the database?_ The result of that query does list "Firmamento" twice for London (which is the expected answer). But further down, it shows rows for London with all the other titles. 

My desired result would have a single row for each of the cities, showing its most "popular" (frequently-purchased) song title.

> What is the role of the nested group?

Maybe I don't understand the semantics of `group`... Coming to SQL from a procedural programming perspective, I had expected that ```group { billing_city, billing_country }``` would "collect up" all the rows that had the same city/country and then "apply" the inner pipeline - whatever is between the `( ... )` - to those rows. 

That's why I expected the inner pipeline to be along the lines of:

```
# using all the rows that have the same city/country
group {t.name} (                  # group the track names together
  aggregate { c = count t.name }  # count 'em
  sort {-c}                       # find the largest count
  take 1                          # and pass along the first row
)
```
My questions:

1. Do I have the correct mental model for `group`? 
2. Should I expect that the table passed into the nested group (from the outer `group { city, country }`)would contain those columns?
3. Should I expect that the rows passed out of the nested group would contain [ city, country, name, c ] as well as any other columns passed in from the outer group?

Thanks

**Comment by max-sixty** [comment]
That makes sense, thanks for clarifying.

I think this should do what you need (simplified for country only), but unfortunately falls prey to https://github.com/PRQL/prql/issues/3111.

```elm
from inv=invoices
join item=invoice_items (==invoice_id)
join t=tracks (item.track_id==t.track_id)

group {inv.billing_country, t.name} (
  aggregate (c=count t.name)
)

group {billing_country} (
  aggregate (
    sort (-c)
    take 1
  )
)
```

So in this case, I don't think we need a nested group, I think we can just have them be serial. Does that make sense to you?

(I realize it's more difficult to discuss when there's a bug that prevents the query from running! Hopefully we'll solve that one soon...)

**Comment by richb-hanover** [comment]
Thanks, that helps. It sounds as if my intuition is correct: `group` is designed to collect lines with identical key_columns, and run them through the inner pipeline. And the [inner pipeline I proposed above](https://github.com/PRQL/prql/issues/3490#issuecomment-1712954214) would do what I want, except that a bug (#3111) causes this not to work presently.

I also agree that the two serial `group`s will do what I want. Is this a general rule - that you can reformulate a query different ways (as with Perl, "There’s more than one way to do it") or is this example just a convenient special case? (I don't understand enough about relational algebra to say...)

New question: I don't see anyplace where it's documented what columns are passed to the inner pipeline, or the columns that will be passed out. More questions, and maybe answering myself...

* If the inner pipeline of a `group` does _not_ contain `aggregate`, all columns are passed through. The inner pipeline could conceivably add or remove columns (with `derive` or `select`) but those statements could be placed outside the `group` with substantially the same effect.

* If the inner pipeline _does_ contain `aggregate`, there's an "implicit select" of the _key\_columns_ and any columns created by `aggregate` - no other columns are passed through. That's because the `aggregate` changes the "shape" of the table, by reducing multiple rows to a single row.  It's not meaningful to talk about values from an individual row anymore - the data from multiple rows has been summarized - so it cannot be passed through un-summarized.

Finally, one last note: the query below gives the results (one song per city) I expected. (The query above doesn't need the second `aggregate`) Thanks as always.

```
# What's the most popular song in each CITY
from inv=invoices
join item=invoice_items (==invoice_id)
join t=tracks (item.track_id==t.track_id)

group {inv.billing_city, t.name} (
  aggregate (c=count t.name)
)

group {inv.billing_city} (
    sort (-c)
    take 1
)
sort inv.billing_city
````

**Comment by aljazerzen** [comment]
Disclaimer: I did not read all of the above.

I think there is a deeper issue with how group is currently implemented and this shows up when you use nested `group`.

As I see it, behavior of group is:
- split the relation into partition using the `by` argument,
- apply the `pipeline` function argument to each of the partitions,
- prefix the partitions with the `by` argument,
- concatenate partitions back together into a single relation.

```elm
from x
select {a, b, c}
group a (
  aggregate {d = sum c}  # this will produce partitions of form [{d = 12.3}]
)
# concatenated back together, result will have form
# [
#   {a = 1, d = 12.3},
#   {a = 2, d = 43.1},
# ]
```

So nesting should work like this:

```elm
from x
select {a, b, c}
group a (
  group b (
    aggregate {d = sum c}  # this will produce partitions of form [{d = 12.3}]
  )
  # concatenated back together, result will have form
  # [
  #   {b = 's', d = 12.3},
  #   {b = 't', d = 43.1},
  # ]
)
# concatenated back together, result will have form
# [
#   {a = 1, b = 's', d = 12.3},
#   {a = 1, b = 't', d = 43.1},
#   {a = 2, b = 's', d = 53.3},
#   {a = 2, b = 't', d = 65.4},
# ]
```

**Comment by richb-hanover** [comment]
> As I see it, behavior of group is:
> 
> 	• split the relation into partition using the by argument,
> 	• apply the pipeline function argument to each of the partitions,
> 	• prefix the partitions with the by argument,
> 	• concatenate partitions back together into a single relation.

@aljazerzen I like this description, but I think the third step is more complex: As I wrote above:

> * If the inner pipeline of a `group` does _not_ contain `aggregate`, all columns are passed through. The inner pipeline could conceivably add or remove columns (with `derive` or `select`) but those statements could be placed outside the `group` with substantially the same effect.
> 
> * If the inner pipeline _does_ contain `aggregate`, there's an "implicit select" of the _key\_columns_ and any columns created by `aggregate` - no other columns are passed through. That's because the `aggregate` changes the "shape" of the table, by reducing multiple rows to a single row.  It's not meaningful to talk about values from an individual row anymore - the data from multiple rows has been summarized - so it cannot be passed through un-summarized.

Other observations:

* if the `group`'s pipeline merely selects a couple columns, the query does not actually split the relation into partitions - it simply passes the rows through, possibly in the original order. In addition, the `by` arguments aren't always passed through. For example:

```
from invoices
group { billing_city, billing_country } (
  select {
    billing_city,
    billing_address, # NOT billing_country
    invoice_id,
    total
  } 
)
```

* Simply sorting the inner pipeline does not change the output - it remains unsorted

```
from invoices
group { billing_city, billing_country } (
  select {
    billing_city,
    billing_address, # NOT billing_country
    invoice_id,
    total
  } 
  sort billing_city
)
```

* However, adding `take 2` gives the expected partition and sorting

```
from invoices
group { billing_city, billing_country } (
  select {
    billing_city,
    billing_address, # NOT billing_country
    invoice_id,
    total
  } 
  sort billing_city
  take 2
)
```
Is this intended? Or is it still related to #3111? Thanks

**Comment by aljazerzen** [comment]
> but I think the third step is more complex

You are right - currently group changes behavior depending on if it contains aggregate or not. Which is against our orthogonality principle.

Could we change it, so it would always prefix `by` columns to the resulting relation?

---

> if the group's pipeline merely selects a couple columns, the query does not actually split the relation into partitions

Right now, this in intended behavior. When the partitions are concatenated back together, they loose all order.

It would probably make more sense if the result would remain partitioned, yes?

Something like this:
```elm
from x
group a (select {b, c})


[
  {a = 4, b = ..., c = ...},
  {a = 4, b = ..., c = ...},
  {a = 9, b = ..., c = ...},
  {a = 1, b = ..., c = ...},
  {a = 1, b = ..., c = ...},
  {a = 1, b = ..., c = ...},
]
```

We could achieve this by compiling to `ORDER BY a`. This would imply that partitions are also guaranteed to be sorted, even though PRQL semantics would not guarantee it.

A bit less performant, but also less misleading would be to compile to `ORDER BY md5(a)`

Let's continue this in #1457

---

> Simply sorting the inner pipeline does not change the output - it remains unsorted

Similarly as above, I assume you'd expect this behavior:

```elm
from x
group a (sort b)


[
  {a = 4, b = 3, c = ...},
  {a = 4, b = 6, c = ...},
  {a = 9, b = 2, c = ...},
  {a = 1, b = 44, c = ...},
  {a = 1, b = 45, c = ...},
  {a = 1, b = 46 c = ...},
]
```

Similarly, this would be accomplished with `ORDER BY a, b` or `ORDER BY md5(a), b`

**Comment by aljazerzen** [comment]
Summary:
- the unexpected behavior can be discussed in the linked issue,
- we the following open question:

Should we change the behavior of group to "always prefix the result with the `by` columns" (not just when the pipeline ends with `aggregate`)?

It would mean that in some cases, columns would get duplicated:
```
from x
select {a, b, c}
group a (take 2)


[
  {a = 4, a = 4, b = ..., c = ...},
  {a = 4, a = 4, b = ..., c = ...},
  {a = 9, a = 9, b = ..., c = ...},
  {a = 1, a = 1, b = ..., c = ...},
  {a = 1, a = 1, b = ..., c = ...},
  {a = 1, a = 1, b = ..., c = ...},
]
```

**Comment by aljazerzen** [comment]
A fix for this problem would be to remove the columns from the partitions, before the function is applied to partitions.

```
from x
select {a, b, c}
group a (
  # here, only columns {b, c} are available

  take 2 # this will produce partitions of form [{b = ..., c = ...}]
)
# concatenated back together, result will have form
# [
#   {a = 1, b = ..., c = ...},
#   {a = 1, b = ..., c = ...},
#   {a = 2, b = ..., c = ...},
# ]
```

Although this is problematic from another standpoint: `group`'s `by` argument might be passed columns that are not in the original relation:

```
from x
select {a, b, c}
group {a + 1} (take 2)
```

Problem: how do you remove `{a + 1}` from `{a, b, c}`?

**Comment by snth** [comment]
> As I see it, behavior of group is:
> 
> split the relation into partition using the by argument,
> apply the pipeline function argument to each of the partitions,
> prefix the partitions with the by argument,
> concatenate partitions back together into a single relation.

I think this description is spot on and this behaviour should be retained/strived for in all cases where possible to retain orthogonality.

**Comment by snth** [comment]
Disclaimer: I also did not read everything above in detail.

> Should we change the behavior of group to "always prefix the result with the by columns" (not just when the pipeline ends with aggregate)?

Yes, but only logically. The implementation to achieve may need to vary to accommodate this (IDK)?

> It would mean that in some cases, columns would get duplicated:
> 
> ```elm
> from x
> select {a, b, c}
> group a (take 2)
> ```
No, the current behaviour (at least in the cases I've seen) does have this property.

Take this example:
```elm
# 3 albums with the longest track
from tracks
group {album_id} (
  sort {-milliseconds}
  take 1
)
sort {-milliseconds}
take 3
```
By virtue of keeping all columns in the resultset you are including the `by` columns and you have still partitioned by them. 

I see no good reason to change that at this point.

**Comment by snth** [comment]
> You are right - currently group changes behavior depending on if it contains aggregate or not. Which is against our orthogonality principle.

My first intuition here is that the rule should probably not be around whether there is an `aggregate` in the `group` pipeline but whether it contains a transform with a column restriction like `aggregate` or `select`.

- if there is a column restriction anywhere in the pipeline then prepend the `by` columns. Repetition of any `by` columns would then be intentional.
- if there is no column restriction then don't prepend the `by` columns.

Haven't thought this through thoroughly but it seems to work on the examples I can think of off the top of my head.

**Comment by snth** [comment]
@snth Note to self: I want to add something on nested `group`s but out of time rn.

**Comment by aljazerzen** [comment]
> if there is a column restriction anywhere in the pipeline then ...

This also makes the behavior of group depend on the function that is applied to the partitions.

**Comment by snth** [comment]
> > if there is a column restriction anywhere in the pipeline then ...
> 
> This also makes the behavior of group depend on the function that is applied to the partitions.

I'm suggesting this as an implementation detail in the generated SQL in order to keep the logical behaviour of PRQL consistent.

I agree that the PRQL behaviour should be independent of the internals of the `group` pipeline.

**Comment by snth** [comment]
> So in this case, I don't think we need a nested group, I think we can just have them be serial. Does that make sense to you?

My gut feeling around this is that this should hold in general and could be an implementation strategy for the compiler. In particular I believe that:
```elm
group {a_1, ..., a_i} (
  transform_1
  group {b_1, ..., b_j} (
    transform_2
  )
  transform_3
)
```
should generally be equivalent to:
```elm
group {a_1, ..., a_i} (
  transform_1
)
group {a_1, ..., a_i, b_1, ..., b_j} (
    transform_2
)
group {a_1, ..., a_i} (
  transform_3
)
```
The only complication I can see is around preservation of ordering if there are `sort` transforms in `transform_1` and `transform_2`. 
Can we prove this or can someone think of a counterexample?

My feeling is that since SQL doesn't have nested GROUP BYs AFAIK, this approach of flattening nested `group`s would allow the compiler to transform such PRQL code into forms that can be easily compiled to SQL.

**Comment by max-sixty** [comment]
Centering on the question of orthogonality above (v good discussion): to summarize — the issue is that "`group` should add back its columns" makes a lot of sense in:

```elm
from artists
select {artist, city, visit_date}
group {city, artist} (
  aggregate (visit_count = count visit_date)
}
```

...because we want `{city, artist, visit_count}`. But does not seem correct in:

```elm
from artists
select {artist, city, visit_date}
group {city, artist} (
  sort visit_date
  aggregate (take 1)
)
```

...because we want `{city, artist, visit_date}` again, not ``{city, artist, city, artist, visit_date}`.

---

I really like [@aljazerzen 's suggestion](https://github.com/PRQL/prql/issues/3490#issuecomment-1715682964) of removing the columns before passing in the group — simple semantics that works for both cases.

> Although this is problematic from another standpoint: `group`'s `by` argument might be passed columns that are not in the original relation:
> `group {a + 1} (take 2)`[^1] [^2]

And sometimes you might want to reference the column:

```
from artists
select {artist, city, visit_date}
group {city, artist} (
  sort visit_date
  aggregate (visit_count = count visit_date)
  select message = f"{artist} visited {city} {visit_count} times"
)
```

So even "remove any columns that are not in the relation" might not be sufficient.

At risk of complicating this — one rule that satisfies all of these examples is:

**Remove `by` columns from the relation within the group from `*` if they are in the input relation, but let them be accessed by name**

I'm not sure how easy it to be completely consistent _in SQL_ though, given we don't have access to the column names. We can make it work for 90% of the cases, but sometimes are we just not going to know whether they're there?

[^1]: One way of thinking of this is that `group (foo + 1)` desugars to `derive _unnamed = foo + 1 | group (_unnamed)`. Or disallowing expressions there.

[^2]: Separately, we could allow assigns as arguments to functions, so someone can pass `group (foo = bar + 1)`

**Comment by aljazerzen** [comment]
> And sometimes you might want to reference the column:

That's true, but I think all of the cases where you'd want to reference the `by` columns from within the group partition function, could be transformed by moving the reference out the function:

```
from artists
select {artist, city, visit_date}
group {city, artist} (
  sort visit_date
  aggregate (visit_count = count visit_date)
)
select message = f"{artist} visited {city} {visit_count} times"
```

----

> Remove by columns from the relation within the group from * if they are in the input relation, but let them be accessed by name

This would solve the problem, but we don't do that anywhere else - it would be quite unprecedented. It would make group partition function dependent on the outer context.

**Comment by max-sixty** [comment]
> That's true, but I think all of the cases where you'd want to reference the `by` columns from within the group partition function, could be transformed by moving the reference out the function:

Yes true, good point.

So does that leave:
1. Remove `by` columns from the relation within the group, if they are in the input relation (i.e. not if there's an expression such as `group (foo + 1)`)
2. Remove `by` columns from the relation within the group + disallow expressions in `by`

---



> > Remove by columns from the relation within the group from * if they are in the input relation, but let them be accessed by name
> 
> This would solve the problem, but we don't do that anywhere else - it would be quite unprecedented. It would make group partition function dependent on the outer context.

Yes, it would be unprecedented. To confirm the original proposal, it doesn't rely on any context from outside the group transform — "if they are in the input relation" is true when there's a simple `ident` in the `by`, and false when there's an expression `group (foo + 1)`.

**Comment by richb-hanover** [comment]
> Remove `by` columns from the relation within the group from `*` if they are in the input relation, but let them be accessed by name

Now it's my turn to confess that I haven't completely understood the foregoing discussion. Let me express my (simple-minded) mental model:

- The `group` transform partitions the table according to its `by` columns (I think that's the name we're using...) 

- `group` passes each partition (with all columns, including the `by` columns) to the inner pipeline 

- The inner pipeline does its business: changing/selecting rows (`select`, `derive`, `take`, `filter`, etc) OR summarizing the rows (with `aggregate`). In either case, the inner pipeline produces a result table.

- `group` ensures that the `by` columns are present in the result table, adding them back if they weren't present in the result table, which gets passed to the next transform.

Is this where the design is heading? If so, it will be easy to explain to newcomers. Many thanks.

**Comment by aljazerzen** [comment]
@richb-hanover This is a good way to describe how `group` works at the moment. The only difference between current behavior and what you wrote is the last point: "group ensures that the by columns are present in the result table". This would be "if the last function in the pipeline is aggregate, by columns are added to result table[^1]".

We are now in the process of discussing a change to this behavior, with the primary goal of making it more orthogonal/independent of the inner function.

[^1]: what you call "table" we call "relation"

**Comment by richb-hanover** [comment]
@aljazerzen , @max-sixty , @snth 

> We are now in the process of discussing a change to this behavior, with the primary goal of making it more orthogonal/independent of the inner function.

I had trouble following all the nuances of the discussion above. Have we settled on a design decision? Would you summarize how it will work going forward? Many thanks.

**Comment by aljazerzen** [comment]
In the last comment, max outlined two options that we have.

I don't have a strong opinion on any of the two, so we could leave it up to what's easier to implement.

**Comment by aljazerzen** [comment]
While working on our type system, I stumbled into this.

The only solution to express the behavior of `group` was to update the behavior of group to:
- remove the `by` columns before applying the pipeline to the partition,
- always re-add `by` columns back into the partition.

To do this, I'll have to pick @max-sixty option 2:

> 2. Remove `by` columns from the relation within the group + disallow expressions in by

In practice these are the changes:

1. `by` columns are not available within the pipeline function:
```
    group {a} (derive c = a + 1)
       ╭─[:4:27]
       │     from x
       |     select {a, b}
     4 │     group {a} (derive c = a + 1)
       │                           ┬
       │                           ╰── Unknown name `a`
    ───╯
```

2. `by` columns cannot be expressions
```
    from x
    select {a, b}
    group {a + 1} (aggregate {sum b})
           ^^^^^
               \__ function `group`, param `by` expected plain identifiers, found `a + 1`
```

3. `by` columns are moved in front of the output relation
```
from x
select {a, b}
group {b} (take 1)

# will have the type of
[{b, a}]
```

**Comment by max-sixty** [comment]
Thanks @aljazerzen 

And to confirm for future travelers — this doesn't represent any loss functionality — it's always possible to add a column before the group. So while this will raise an error:

```elm
from x
select {a, b}
group {a + 1} (aggregate {sum b})
```

...we can just break that out:

```elm
from x
select {a, b}
derive foo = a + 1
group {foo} (aggregate {sum b})
```

---

In the future possibly we could allow an assignment in there:

```elm
from x
select {a, b}
group {foo = a + 1} (aggregate {sum b})
```

**Comment by richb-hanover** [comment]
@aljazerzen @max-sixty I still have questions about the query from the original item of this issue.

Where is the best place to continue the discussion? (At this point, it's not a bug report - perhaps in the Discussions?) Many thanks.

**Comment by max-sixty** [comment]
Discussions is fine or adding here is OK!

**Comment by richb-hanover** [comment]
Since this is closed, moved discussion to #3832
